app/services/cache/MongoCacheClient.scala

```diff
@@ -114,53 +114,58 @@ class MongoCacheClient(appConfig: AppConfig, db: () => DefaultDB, applicationCry
   /**
     * Inserts data into the cache with the specified key. If the data does not exist, it will be created.
     */
-  def createOrUpdate[T](id: String, data: T, key: String)(implicit writes: Writes[T]): Future[Cache] = {
+  def createOrUpdateWithCacheMiss[T](id: String, newId: String, data: T, key: String)(implicit writes: Writes[T]): Future[Cache] = {
     val jsonData = if (appConfig.mongoEncryptionEnabled) {
       val jsonEncryptor = new JsonEncryptor[T]()
       Json.toJson(Protected(data))(jsonEncryptor)
     } else {
       Json.toJson(data)
     }
 
-    fetchAll(id) flatMap { maybeCache =>
-      val cache = maybeCache.getOrElse(Cache(id, Map.empty))
+    fetchAll(newId, false) flatMap { maybeNewCache =>
+      fetchAll(id, true) flatMap { maybeCache =>
+        val cache = maybeNewCache.getOrElse(maybeCache.getOrElse(Cache(newId, Map.empty)))
 
-      val updatedCache = cache.copy(
-        data = cache.data + (key -> jsonData),
-        lastUpdated = DateTime.now(DateTimeZone.UTC)
-      )
+        val updatedCache: Cache = cache.copy(
+          id = newId,
+          data = cache.data + (key -> jsonData),
+          lastUpdated = DateTime.now(DateTimeZone.UTC)
+        )
 
-      val document = Json.toJson(updatedCache)
-      val modifier = BSONDocument("$set" -> document)
+        val document = Json.toJson(updatedCache)
+        val modifier = BSONDocument("$set" -> document)
 
-      collection.update(bsonIdQuery(id), modifier, upsert = true) map { _ => updatedCache }
+        collection.update(bsonIdQuery(newId), modifier, upsert = true) map { _ => updatedCache }
+      }
     }
   }
 
   /**
     * Removes the item with the specified key from the cache
     */
-  def removeByKey[T](id: String, key: String)(implicit writes: Writes[T]): Future[Cache] = {
+  def removeByKeyWithCacheMiss[T](id: String, newId: String, key: String)(implicit writes: Writes[T]): Future[Cache] = {
 
-    fetchAll(id) flatMap { maybeCache =>
-      val cache = maybeCache.getOrElse(Cache(id, Map.empty))
+    fetchAll(newId, false) flatMap { maybeNewCache =>
```

Comment:
Could use a for comprehension, rather than a nesting

Reply:
I'd rather leave as it is, because can't see potential benefit of change